### PR TITLE
working on genome wide linkage analysis

### PR DESCRIPTION
**Pull request by changebio**

### CODE REVIEW

**Comment by gaow** [comment]
@changebio This is a large patch and difficult for me to review ... I will try my best. Such a large PR is also hard for yourself to track the evolution of your code. As established, more frequent PR with smaller patches in the future is more reasonable to me.

File: SEQLinkage/Core.py [hunk 1]
```diff
@@ -272,11 +296,25 @@ def extract_vcf_with_anno(self,data):
...
                         for person, g in zip(data.families[k], gs):
                             data[person].append(g if maf<0.5 else self.reverse_genotypes(g))
-            data.variants.append([self.vcf.GetChrom(), self.vcf.GetPosition(), self.name]) #remove maf
+            data.variants.append([self.vcf.GetVariantID(), self.vcf.GetPosition(), self.name]) #remove maf
             #print(i,varmafs.shape,self.chrom, self.startpos, self.endpos, self.name,self.vcf.GetPosition())
             varIdx += 1
         return varIdx
 
+    def check_gs(self,gs):
+        '''skip monomorphic variants and signton variants in a family'''
```
Comment (gaow) [inline SEQLinkage/Core.py:305]: "singleton" is a typo.

Comment (gaow) [reply SEQLinkage/Core.py:305]: Make sure that missing genotype is properly handled here too.

File: SEQLinkage/Core.py [hunk 2]
```diff
@@ -666,8 +687,52 @@ def run(self):
...
+            items.append(item)
+    return items
+
+haplotyper = cstatgen.HaplotypingEngine(verbose = env.debug)
+def phasing_haps(chrom,item,fvar,fgeno):
+    item_varnames, positions, item_mafs = fvar
+    try:
+        item_haplotypes = haplotyper.Execute(chrom, item_varnames, positions, fgeno)[0]
+    except:
+        print(item,"fail to phase haplotypes")
```
Comment (gaow) [inline SEQLinkage/Core.py:706]: please use `env.logger.error` for this type of messages

File: SEQLinkage/Core.py [hunk 3]
```diff
@@ -351,32 +388,12 @@ def __Haplotype(self, data):
...
-            # haplotyping
-            with env.lock:
-                if not env.prephased:
-                    tmp_log_output=env.tmp_log + str(os.getpid()) + '.log'
-                    item_haplotypes = self.haplotyper.Execute(data.chrom, item_varnames, positions, data.getFamSamples(item), self.rsq, tmp_log_output)[0]
-                else:
-                    item_haplotypes = self.__PedToHaplotype(data.getFamSamples(item))
-                item_haplotypes = np.array(item_haplotypes)
-            #env.dtest[self.name]['hapimp'][item].append(item_haplotypes)#test line
+        for item,item_varnames,item_mafs,item_haplotypes in inputs:
```
Comment (gaow) [inline SEQLinkage/Core.py:396]: lots of code are deleted here. Can you tell me why they are removed (and make sure it's fine to do so?)

File: SEQLinkage/Core.py [hunk 4]
```diff
@@ -666,8 +687,52 @@ def run(self):
...
+    item_varnames, positions, item_mafs = fvar
+    try:
+        item_haplotypes = haplotyper.Execute(chrom, item_varnames, positions, fgeno)[0]
+    except:
+        print(item,"fail to phase haplotypes")
+        item_haplotypes = []
+    item_haplotypes = np.array(item_haplotypes)
+    return item,item_varnames,item_mafs,item_haplotypes
+
+def test(region,data,extractor,maker,writer):
```
Comment (gaow) [inline SEQLinkage/Core.py:711]: this is a test function and should not be in the code. If you want to have test functions can you make those separate python scripts? eg something like test.py that you can tell from the name that it is not part of the core features?

File: SEQLinkage/Core.py [hunk 5]
```diff
@@ -692,7 +757,61 @@ def run_each_region(regions,data,extractor,maker,writer):
             with env.success_counter.get_lock():
                 env.success_counter.value += 1
             results[region[3]]=maker.dtest[region[3]]
-    env.log('write to pickle')
-    import pickle
-    with open(os.path.join(env.tmp_cache,env.output+'.pickle'), 'wb') as handle:
-        pickle.dump(results, handle, protocol=pickle.HIGHEST_PROTOCOL)
\ No newline at end of file
+            if len(results)==25:
```
Comment (gaow) [inline SEQLinkage/Core.py:760]: we are using this mechanism to cache and prevent too much data in the memory? Let's make it a function parameter "cache_size = 25" with default 25.

File: SEQLinkage/Core.py [hunk 6]
```diff
@@ -692,7 +757,61 @@ def run_each_region(regions,data,extractor,maker,writer):
...
+                with open(os.path.join(env.tmp_cache,env.output+str(i)+'.pickle'), 'wb') as handle:
+                    pickle.dump(results, handle, protocol=pickle.HIGHEST_PROTOCOL)
+                results = {}
+                i +=1
+    env.log('write to pickle: '+os.path.join(env.tmp_cache,env.output+str(i)+'.pickle')+',Gene number:'+str(len(results))+',Time:'+str((time.perf_counter()-start)/3600))
+    with open(os.path.join(env.tmp_cache,env.output+str(i)+'.pickle'), 'wb') as handle:
+        pickle.dump(results, handle, protocol=pickle.HIGHEST_PROTOCOL)
+        results = {}
+
+def run_each_region_genotypes(regions,data,extractor,maker,writer):
```
Comment (gaow) [inline SEQLinkage/Core.py:772]: this is a new function. Please document in comment string what this functoin does.

File: SEQLinkage/Core.py [hunk 7]
```diff
@@ -108,6 +113,23 @@ def load_fam_info(self,fam_pop_file):
     def load_ind_samples(self,ind_sample_file):
         pass
 
+    def get_regions(self,step=1000):
```
Comment (gaow) [inline SEQLinkage/Core.py:116]: this is a new function. Please document what it is for.

File: SEQLinkage/linkage.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
+# AUTOGENERATED! DO NOT EDIT! File to edit: nbs/linkage_analysis.ipynb (unless otherwise specified).
```
Comment (gaow) [inline SEQLinkage/linkage.py:1]: is this file supposed to replace the `run_linkage.py`

File: SEQLinkage/Utils.py [hunk 9]
```diff
@@ -120,7 +120,8 @@ def __del__(self):
                     remove_tree(os.path.join(where, fn))
                 except:
                     pass
-        tmp = LockedTempDir(tempfile.mkdtemp(prefix='{}_tmp_'.format(self.proj), dir = where))
+        #tmp = LockedTempDir(tempfile.mkdtemp(prefix='{}_tmp_'.format(self.proj), dir = where))
```
Comment (gaow) [inline SEQLinkage/Utils.py:123]: why do we have to comment this out? I think it is safer to create project specific cache folders?

File: csg.yml [hunk 10]
```diff
@@ -0,0 +1,52 @@
+hosts:
```
Comment (gaow) [inline csg.yml:1]: this file should not be added to SEQLinkage repository. Please remove. You can put it in alzheimers-family analysis repo if you want.

File: linkagetest.sh [hunk 11]
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
```
Comment (gaow) [inline linkagetest.sh:1]: remove this file. Instead, let's start a unit test folder. We did not have unit test before but we should have it now. Let's do that using Python's unittest module. Tests can be as simple as using `os.system` to run SEQLinkage without errors. But the test data should be small enough to run on a local machine and quickly produce results.

File: monitor.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
```
Comment (gaow) [inline monitor.py:1]: remove this file. It should be in your own utility collection if you find this script useful.

File: SEQLinkage/linkage.py (hunk 8 above)
Comment (changebio) [reply SEQLinkage/linkage.py:1]: Yes. This is the core codes for likelihood calculation.

File: SEQLinkage/Core.py (hunk 3 above)
Comment (changebio) [reply SEQLinkage/Core.py:396]: I added parallelization to phasing haplotypes and reorganized the code.

File: SEQLinkage/Core.py (hunk 1 above)
Comment (changebio) [reply SEQLinkage/Core.py]: updated

File: SEQLinkage/Core.py (hunk 2 above)
Comment (changebio) [reply SEQLinkage/Core.py]: updated

File: SEQLinkage/Core.py (hunk 4 above)
Comment (changebio) [reply SEQLinkage/Core.py]: removed

File: SEQLinkage/Core.py (hunk 5 above)
Comment (changebio) [reply SEQLinkage/Core.py:741]: will do late.

File: SEQLinkage/Core.py (hunk 6 above)
Comment (changebio) [reply SEQLinkage/Core.py:753]: add doc

File: SEQLinkage/Core.py (hunk 7 above)
Comment (changebio) [reply SEQLinkage/Core.py:116]: add doc
This is a function to separate whole genome to regions with 1000 variants if annotation file provide and no reference gene file provide. you can think each region is a "gene".

File: csg.yml (hunk 10 above)
Comment (changebio) [reply csg.yml]: removed

File: linkagetest.sh (hunk 11 above)
Comment (changebio) [reply linkagetest.sh:1]: removed

File: SEQLinkage/Utils.py (hunk 9 above)
Comment (changebio) [reply SEQLinkage/Utils.py:123]: The tmp folder is different each time we run it, which makes no way to using the results from last run. The cache folder is still there, but with specific name.

**Comment by changebio** [comment]
The CI build is a feature of nbdev. The check failing is related with cstatgen. It is because cstatgen can not install automatically by pip.

File: SEQLinkage/Core.py (hunk 5 above)
Comment (changebio) [reply SEQLinkage/Core.py]: updated

**Comment by changebio** [comment]
please merge the PR.

**Comment by gaow** [comment]
@changebio please clean up the `nbs` folder and move all analysis code relevant to the current AD family project to the AD family analysis repository. You can stick to the existing convention of `yyyymmdd_content.ipynb` for the filenames of  the notebook. The SEQLinkage package should only contain source code and unit tests.

**Comment by gaow** [comment]
@changebio are you sure `nbs` folder is completely cleaned up? I see still APOE analysis. Also not all notebooks are ordered with the numerical prefix convention. there are also random R and Python scripts to be cleaned up.

**Pull request merged**